sys/drv/simplebus.c

```diff
@@ -0,0 +1,199 @@
+#define KL_LOG KL_DEV
+#include <sys/bus.h>
+#include <sys/device.h>
+#include <sys/errno.h>
+#include <sys/fdt.h>
+#include <sys/interrupt.h>
+#include <sys/klog.h>
+#include <sys/libkern.h>
+
+static int sb_get_region(device_t *dev, fdt_mem_reg_t *mrs, size_t *cntp) {
+  phandle_t node = dev->node;
+
+  int addr_cells, size_cells;
+  int err = FDT_addrsize_cells(FDT_parent(node), &addr_cells, &size_cells);
+  if (err)
+    return err;
+
+  int tuple_cells = addr_cells + size_cells;
+  size_t tuple_size = sizeof(pcell_t) * tuple_cells;
+  pcell_t *tuples;
+
+  ssize_t ntuples =
+    FDT_getprop_alloc_multi(node, "reg", tuple_size, (void **)&tuples);
+  if (ntuples == -1)
+    return ENXIO;
+
+  if (ntuples > FDT_MAX_REG_TUPLES) {
+    err = ERANGE;
+    goto end;
+  }
+
+  for (int i = 0; i < ntuples; i++) {
+    pcell_t *tuple = &tuples[i * tuple_cells];
+    fdt_mem_reg_t *mr = &mrs[i];
+    if ((err = FDT_data_to_res(tuple, addr_cells, size_cells, &mr->addr,
+                               &mrs->size)))
+      goto end;
+  }
+
+end:
+  *cntp = ntuples;
+  FDT_free(tuples);
+  return err;
+}
+
+static int sb_get_interrupts(device_t *dev, fdt_intr_t *intrs, size_t *cntp) {
+  phandle_t node = dev->node;
+  phandle_t iparent = dev->pic->node;
+
+  int icells;
+  int err = FDT_intr_cells(iparent, &icells);
+  if (err)
+    return err;
+
+  pcell_t *tuples;
+  int ntuples =
+    FDT_getencprop_alloc_multi(node, "interrupts", icells, (void **)&tuples);
+  if (ntuples < 0)
```

Comment:
Error is only on `-1`. Am I missing something?

Reply:
You're right. I changed it to ` == -1`.